### PR TITLE
Add sbomnix to current Jenkins pipeline

### PR DESCRIPTION
**Pull request by mnokka-unikie**

### CODE REVIEW

File: ghaf-build-pipeline.groovy [hunk 1]
```diff
@@ -75,29 +75,79 @@ pipeline {
...
+                  def buildName = build[0]
+                  def outputName = build[1]
+                  sh "nix build -L .${buildName} -o ${outputName}"
+
+                  def sbomnixResult = sh script: "nix run github:tiiuae/sbomnix#sbomnix -- .${buildName} --csv ${outputName}.csv --cdx ${outputName}.cdx.json --spdx ${outputName}.spdx.json", returnStatus: true
+                  if (sbomnixResult != 0) {
+                       currentBuild.result = 'UNSTABLE'
+                       echo "The sbomnix command failed. Setting build status to UNSTABLE."
+                    } else {
+                        echo "The sbomnix command succeeded. Carry on as normally"
```
Comment (henrirosten) [inline ghaf-build-pipeline.groovy:105]: Why not have the sbomnix invocations as a separate step after the build steps?
I would prefer keeping the pipeline script as simple as possible, even at the expense of few more keystrokes.

Also, I believe we will not want to run sbomnix (let alone some other tools we'll later add to this pipeline) against all targets we'll run the build targets for.

**Review by henrirosten — COMMENTED** [review]
In addition to the below comment
- please check the script format (preferably with a tool of somesort). As it now is, the indentation and formatting is a mess.
- please squash any commits that don't need to be kept in the history.
- please provide a short description as to what and why was changed.

File: ghaf-build-pipeline.groovy [hunk 2]
```diff
@@ -75,29 +75,79 @@ pipeline {
...
-              sh 'nix build -L .#packages.x86_64-linux.lenovo-x1-carbon-gen11-debug             -o result-lenovo-x1-carbon-gen11-debug'
-              sh 'nix build -L .#packages.riscv64-linux.microchip-icicle-kit-debug              -o result-microchip-icicle-kit-debug'
-              sh 'nix build -L .#packages.x86_64-linux.doc                                      -o result-doc'
-            }
-          }
+
+              script {
+
+
+               def builds= [
```
Comment (mkaapu) [inline ghaf-build-pipeline.groovy:86]: You can implement this with matrix or parallel stages of declarative pipeline:
https://www.jenkins.io/doc/book/pipeline/syntax/#parallel
If we can keep Jenkinsfiles simple with as less custom scripting as possible, they stay more readable and will be easier to maintain.

**Comment by mkaapu** [comment]
No need to merge this to main.
I could scrap ideas from this PR and add them to refactoring work which is still under work.

SBOM generation has been also tested here:
https://github.com/tiiuae/ghaf-jenkins-pipeline/blob/WIP-PoC-matrix/Jenkinsfile
Jenkinsfile from the WIP-PoC branch will be taken into use in our ghaf-infra dev environment.

**Review by mkaapu — CHANGES_REQUESTED** [review]

File: ghaf-build-pipeline.groovy (hunk 2 above)
Comment (mnokka-unikie) [reply ghaf-build-pipeline.groovy]: Simplified with parallel stages for builds and sbomnix usage. Used parallel operations to speed artifacts creations

**Review by mkaapu — CHANGES_REQUESTED** [review]

File: ghaf-build-pipeline.groovy [hunk 3]
```diff
@@ -95,12 +99,55 @@ pipeline {
...
     }
   }
-  post {
+
+post {
     always {
-      archiveArtifacts allowEmptyArchive: true, artifacts: 'ghaf/result-*/**'
+        script {
+            def directoryPath = 'ghaf'
+            def artifactsToArchive = sh(script: "find ${directoryPath} -type f -name 'result-*'", returnStdout: true).trim().split('\n')
```
[the remark this answers is not in the document]
Comment (henrirosten) [reply ghaf-build-pipeline.groovy:137]: The artifacts archive will soon change with https://ssrc.atlassian.net/browse/SP-4255, so I would not spend too much time improving the 'old' way of archiving artifacts. I would be fine with dropping this part completely from the PR.

File: ghaf-build-pipeline.groovy [hunk 4]
```diff
@@ -95,12 +99,55 @@ pipeline {
...
+                            sh 'nix run github:tiiuae/sbomnix#sbomnix -- .#packages.aarch64-linux.nvidia-jetson-orin-agx-debug --csv result-aarch64-jetson-orin-agx-debug.csv --cdx result-aarch64-jetson-orin-agx-debug.cdx.json --spdx result-aarch64-jetson-orin-agx-debug.spdx.json'
+                            sh 'nix run github:tiiuae/sbomnix#sbomnix -- .#packages.aarch64-linux.nvidia-jetson-orin-nx-debug  --csv result-aarch64-jetson-orin-nx-debug.csv --cdx result-aarch64-jetson-orin-nx-debug.cdx.json --spdx result-aarch64-jetson-orin-nx-debug.json'
+                            sh 'nix run github:tiiuae/sbomnix#sbomnix -- .#packages.aarch64-linux.imx8qm-mek-debug   --csv result-aarch64-imx8qm-mek-debug.csv --cdx result-aarch64-imx8qm-mek-debug.cdx.json --spdx result-aarch64-imx8qm-mek-debug.spdx.json'
+                            sh 'nix run github:tiiuae/sbomnix#sbomnix -- .#packages.aarch64-linux.doc --csv result-aarch64-doc.csv --cdx result-aarch64-doc.cdx.json --spdx result-aarch64-doc.spdx.json'
+                        }
+                    }
+                }
+            }
+        }
       }
```
Comment (henrirosten) [inline ghaf-build-pipeline.groovy:129]: Are you using indentation of 3 spaces here? Please keep the original indentation instead.

File: ghaf-build-pipeline.groovy [hunk 5]
```diff
@@ -95,12 +99,55 @@ pipeline {
...
+            artifactsToArchive.each { artifact ->
+                parallelTasks["Archive ${artifact}"] = {
+                    echo "Archiving artifact: ${artifact}"
+                    archiveArtifacts allowEmptyArchive: true, artifacts: "${artifact}"
+                }
+            }
+            // Execute the archiving tasks in parallel
+            parallel parallelTasks
+            echo 'Artifact archiving completed.'
+        }
```
Comment (henrirosten) [inline ghaf-build-pipeline.groovy:150]: Same thing here, please fix the indentation

File: ghaf-build-pipeline.groovy (hunk 3 above)
[the remark this answers is not in the document]
Comment (mnokka-unikie) [reply ghaf-build-pipeline.groovy]: returned old way with added verbal operation whats being created

File: ghaf-build-pipeline.groovy (hunk 4 above)
Comment (mnokka-unikie) [reply ghaf-build-pipeline.groovy]: gone and checked 2 space steps

File: ghaf-build-pipeline.groovy (hunk 5 above)
Comment (mnokka-unikie) [reply ghaf-build-pipeline.groovy]: gone and checked 2 space steps

**Pull request closed without merging**